File: test/messaging_pricing_test.dart
import 'package:cloud_functions/cloud_functions.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:pasella/services/dynamic_pricing_service.dart';

void main() {
  const valid = <String, dynamic>{
    'schemaVersion': 1,
    'currency': 'ZAR',
    'smsCustomerMinor': 174,
    'smsPaymentMinor': 174,
    'whatsappUtilityMinor': 20,
    'whatsappPromotionMinor': 100,
  };

  test('parses exact integer-cent messaging pricing', () {
    final pricing = MessagingPricingSnapshotV1.fromMap(valid);
    expect(pricing.smsCustomerMinor, 174);
    expect(pricing.smsPaymentMinor, 174);
    expect(pricing.whatsappUtilityMinor, 20);
    expect(pricing.whatsappPromotionMinor, 100);
  });

  for (final key in const [
    'smsCustomerMinor',
    'smsPaymentMinor',
    'whatsappUtilityMinor',
    'whatsappPromotionMinor',
  ]) {
    test('fails closed when $key is missing or zero', () {
      final missing = Map<String, dynamic>.from(valid)..remove(key);
      final zero = Map<String, dynamic>.from(valid)..[key] = 0;
      expect(
        () => MessagingPricingSnapshotV1.fromMap(missing),
        throwsA(isA<MessagingPricingUnavailable>()),
      );
      expect(
        () => MessagingPricingSnapshotV1.fromMap(zero),
        throwsA(isA<MessagingPricingUnavailable>()),
      );
    });
  }

  test('rejects an unexpected schema or currency', () {
    expect(
      () => MessagingPricingSnapshotV1.fromMap({
        ...valid,
        'schemaVersion': 2,
      }),
      throwsA(isA<MessagingPricingUnavailable>()),
    );
    expect(
      () => MessagingPricingSnapshotV1.fromMap({
        ...valid,
        'currency': 'USD',
      }),
      throwsA(isA<MessagingPricingUnavailable>()),
    );
  });

  test('does not round invalid or non-finite monetary values into a price', () {
    for (final invalid in [174.9, double.nan, double.infinity, -1]) {
      expect(
        () => MessagingPricingSnapshotV1.fromMap({
          ...valid,
          'smsCustomerMinor': invalid,
        }),
        throwsA(isA<MessagingPricingUnavailable>()),
      );
    }
  });

  test(
      'pricing request uses the selected shop and preserves safe App Check reason',
      () async {
    String? requestedShop;
    var attempts = 0;
    await expectLater(
      DynamicPricingService.loadSnapshot(
        storeId: 'test-shop',
        refreshCredentials: () async {},
        retryDelay: (_) async {},
        loader: (shop) async {
          attempts++;
          requestedShop = shop;
          throw FirebaseFunctionsException(
            code: 'failed-precondition',
            message: 'App verification is required.',
            details: const {'private': 'must never be exposed'},
          );
        },
      ),
      throwsA(isA<MessagingPricingUnavailable>()
          .having((error) => error.code, 'code', 'app-check-required')
          .having((error) => error.toString(), 'safe message',
              isNot(contains('must never be exposed')))),
    );
    expect(requestedShop, 'test-shop');
    expect(attempts, 2);
  });

  test('transient pricing failure refreshes credentials and retries once',
      () async {
    var attempts = 0;
    var refreshes = 0;
    var delays = 0;
    Future<Map<String, dynamic>> load(String _) async {
      attempts++;
      if (attempts == 1) {
        throw FirebaseFunctionsException(
            code: 'unavailable', message: 'Connection unavailable');
      }
      return valid;
    }

    final snapshot = await DynamicPricingService.loadSnapshot(
      loader: load,
      storeId: 'demo',
      refreshCredentials: () async => refreshes++,
      retryDelay: (_) async => delays++,
    );
    expect(snapshot.smsCustomerMinor, valid['smsCustomerMinor']);
    expect(attempts, 2);
    expect(refreshes, 1);
    expect(delays, 1);
  });

  test('invalid pricing fails closed without retry or credential refresh',
      () async {
    var attempts = 0;
    var refreshes = 0;
    await expectLater(
      DynamicPricingService.loadSnapshot(
        storeId: 'demo',
        loader: (_) async {
          attempts++;
          return {...valid, 'smsPaymentMinor': 0};
        },
        refreshCredentials: () async => refreshes++,
        retryDelay: (_) async {},
      ),
      throwsA(
        isA<MessagingPricingUnavailable>().having(
          (error) => error.code,
          'code',
          'invalid-pricing-response',
        ),
      ),
    );
    expect(attempts, 1);
    expect(refreshes, 0);
  });
}

File: functions/test/messaging_pricing.test.mjs
import assert from "node:assert/strict";
import test from "node:test";

import {
  loadMessagingPricingSnapshotV1,
  MessagingPricingUnavailableError,
  messagingPricingSnapshotV1,
} from "../lib/services/dynamic_pricing_service.js";

const values = {
  USD_SMS_REMINDER_PRICE: "0.0757",
  USD_SMS_PAYMENT_PRICE: "0.0757",
  USD_WHATSAPP_UTILITY_PRICE: "0.0076",
  USD_WHATSAPP_PROMOTIONAL_PRICE: "0.0379",
  MARKUP_SMS_PERCENTAGE: "26",
  MARKUP_WHATSAPP_PERCENTAGE: "45",
  MARKUP_PROMOTIONAL_PERCENTAGE: "45",
  USD_ZAR_EXCHANGE_RATE: "18.25",
};

function parameters(source = values) {
  return Object.fromEntries(
    Object.entries(source).map(([key, value]) => [
      key,
      { defaultValue: { value } },
    ]),
  );
}

test("messaging pricing returns exact integer-cent production economics", () => {
  assert.deepEqual(messagingPricingSnapshotV1({ parameters: parameters() }), {
    schemaVersion: 1,
    currency: "ZAR",
    smsCustomerMinor: 174,
    smsPaymentMinor: 174,
    whatsappUtilityMinor: 20,
    whatsappPromotionMinor: 100,
  });
});

test("messaging pricing reads Firebase parameter groups", () => {
  const snapshot = messagingPricingSnapshotV1({
    parameters: {},
    parameterGroups: {
      messaging: { parameters: parameters() },
    },
  });
  assert.equal(snapshot.smsCustomerMinor, 174);
  assert.equal(snapshot.whatsappUtilityMinor, 20);
});

for (const [name, override] of [
  ["missing", undefined],
  ["zero", "0"],
  ["malformed", "free"],
]) {
  test(`messaging pricing fails closed for a ${name} required rate`, () => {
    const source = { ...values };
    if (override === undefined) delete source.USD_WHATSAPP_UTILITY_PRICE;
    else source.USD_WHATSAPP_UTILITY_PRICE = override;
    assert.throws(
      () => messagingPricingSnapshotV1({ parameters: parameters(source) }),
      MessagingPricingUnavailableError,
    );
  });
}

test("messaging pricing retries one transient template read", async () => {
  let attempts = 0;
  let delays = 0;
  const snapshot = await loadMessagingPricingSnapshotV1(
    async () => {
      attempts += 1;
      if (attempts === 1) throw new Error("temporary provider failure");
      return { parameters: parameters() };
    },
    async () => {
      delays += 1;
    },
  );

  assert.equal(attempts, 2);
  assert.equal(delays, 1);
  assert.equal(snapshot.smsPaymentMinor, 174);
});

test("messaging pricing does not retry an invalid fetched template", async () => {
  let attempts = 0;
  await assert.rejects(
    loadMessagingPricingSnapshotV1(
      async () => {
        attempts += 1;
        return {
          parameters: parameters({ ...values, USD_SMS_PAYMENT_PRICE: "0" }),
        };
      },
      async () => assert.fail("invalid pricing must not be retried"),
    ),
    MessagingPricingUnavailableError,
  );
  assert.equal(attempts, 1);
});

test("messaging pricing stops after one failed provider retry", async () => {
  let attempts = 0;
  let delays = 0;
  await assert.rejects(
    loadMessagingPricingSnapshotV1(
      async () => {
        attempts += 1;
        throw new Error("provider unavailable");
      },
      async () => {
        delays += 1;
      },
    ),
    /provider unavailable/,
  );
  assert.equal(attempts, 2);
  assert.equal(delays, 1);
});
